fix(upload): return type on POST is now string

lyager opened the pull request:
A POST to a webserver can not always be expected to be a JSON response.

Success is now determined by the HTTP return code.

Upon success the body content is returned as a string.

lyager committed:
fix(upload): return type on POST is now string

A POST to a webserver can not always be expected to be a JSON response.

Success is now determined by the HTTP return code.

Upon success the body content is returned as a string.

lyager commented:
This might be more of a suggestion - not all web services receiving a post returns JSON. Propose to leave result processing to the end-user

FabianLars commented:
Hmm, you're right with that, however we actually ignore the return value on the js side so we may as well return `Result<()>` from the command instead so we can get rid of both res.json() and res.text() and really only use the status.

For that status error i'd prefer if you could add a new Error enum variant instead of using ContentLength

Lastly, it would be much appreciated if you could apply the same to the download command after making those changes :)

P.S. Thank you for contributing!

lyager committed:
feat: add content-length on POST

Not all embedded devices are acceptable to receiving unspecified amounts
of data. Appending the content-length up front helps this devices
succeed.

lyager commented:
Thanks for the feedback @FabianLars I somehow missed your response. I'll give it another shot hopefully soon

lyager committed:
fix: return values unified

The return values was not used.

On POST the HTTP error code is returned as an enum.

lyager commented:
@FabianLars Ok, I think that was all of your suggestions :) What do you think?

lyager commented:
.. my default linter formater apperantly is different that `cargo fmt` - I fixed it. Sorry.

lucasfernog commented on plugins/upload/src/lib.rs:
```diff
@@ -108,8 +114,10 @@ async fn upload<R: Runtime>(
     let response = request.send().await?;
-
-    response.json().await.map_err(Into::into)
+    if response.status().is_success() {
+        return response.text().await.map_err(Error::from);
+    }
+    Err(Error::HttpErrorCode(response.status().as_u16()))
```
```suggestion
    if response.status().is_success() {
        response.text().await.map_err(Into::into)
    } else {
        Err(Error::HttpErrorCode(response.status().as_u16()))
    }
```

FabianLars commented:
There's still this part open from above:
> Hmm, you're right with that, however we actually ignore the return value on the js side so we may as well return Result<()> from the command instead so we can get rid of both res.json() and res.text() and really only use the status.

So if there's any value in the response body we'll also have to fix the js side to actually forward it to the user.

lyager committed:
fix: upload, return value as string

lyager commented:
@FabianLars You're right. It should be fixed now.

I also added @lucasfernog suggestion.

I'll do some extensive test, but I think this should finally be it 😅

lucasfernog commented on plugins/upload/src/lib.rs:
```diff
@@ -29,6 +29,8 @@ pub enum Error {
     Request(#[from] reqwest::Error),
     #[error("{0}")]
     ContentLength(String),
+    #[error("{0}")]
+    HttpErrorCode(u16),
```
```suggestion
    #[error("request failed with status code {0}: {1}")]
    HttpErrorCode(u16, String),
```

lucasfernog commented on plugins/upload/src/lib.rs:
```diff
@@ -108,8 +114,11 @@ async fn upload<R: Runtime>(
+    if response.status().is_success() {
+        response.text().await.map_err(Into::into)
+    } else {
+        Err(Error::HttpErrorCode(response.status().as_u16()))
```
```suggestion
        Err(Error::HttpErrorCode(
            response.status().as_u16(),
            response.text().await.unwrap_or_default(),
        ))
```

lucasfernog reviewed (dismissed):
We also need a change file (a markdown file under the .changes folder) like this:
```
---
"upload": patch
"upload-js": patch
---

Return the upload response as a string and error out if the status code is not within 200-299.
```

lucasfernog commented:
Tried to push changes to your repo but i'm not allowed to :(

lyager committed:
Update plugins/upload/src/lib.rs

Co-authored-by: Lucas Fernandes Nogueira <lucas@tauri.app>

lyager committed:
Update plugins/upload/src/lib.rs

Co-authored-by: Lucas Fernandes Nogueira <lucas@tauri.app>

lyager commented:
@lucasfernog Thank you for your help, there are clearly some corners in the repo that are unknown to me. I've accepted you proposals in the Github UI.

lucasfernog commented:
@lyager please also push the change file i mentioned above :)

lyager committed:
fix: added covector changelog file

lyager commented:
Thanks for the patience @lucasfernog , I've rebased and added the covector file you mentioned.

lucasfernog commented:
idk why but the diff is crazy right now :( something went wrong with your rebase

lyager commented:
@lucasfernog I've rebased against v1, now only 3 files have changed.. Sorry about that, and thank you for your patience.
